lib/runtime/procs/read.js

```diff
@@ -0,0 +1,282 @@
+        else if (this.from.lt(this.program.now) || !this.to.isEnd()) {
+            this.readFrom = this.from;
+            if (this.to.isEnd()) {
+                // XXX/demmer need to handle the case where now - lag > from
+                this.readTo = this.program.now.subtract(this.lag);
+            } else {
+                this.readTo = this.to;
+            }
+            this.logger.debug('starting historical read at', this.program.now.toJSON(),
+                              'from', this.readFrom.toJSON(), 'to', this.readTo.toJSON());
+            this.run();
+        }
+        // For fully "live", the first chunk of points to read is from
+        // [now ... now + readEvery).
+        //
+        // So before we can actually read this initial set of points are
+        // read, we need to wait for both the readEvery period as well as
+        // the lag time to pass.
+        else {
+            this.readFrom = this.program.now;
+            this.readTo = this.program.now.add(this.readEvery);
+            var firstRead = this.program.now.add(this.readEvery).add(this.lag);
+            this.logger.debug('scheduling first live read at', firstRead.toJSON());
+            this.program.scheduler.schedule(firstRead.unixms(), _.bind(this.run, this));
+        }
+    },
+
+    // Kick off a read iteration from the adapter.
+    run: function() {
+        if (this.reading) {
+            throw new Error('run should not be called while reading');
+        }
+        this.logger.debug('read from', this.readFrom.toJSON(),
+                          'to', this.readTo.toJSON(),
+                          'limit', this.queueSize, 'state', this.readState);
+
+        this.reading = Promise.try(() => {
+            return this.adapter.read(this.readFrom, this.readTo, this.queueSize, this.readState);
+        })
+        .then((result) => {
+            this.reading = null;
+
+            if (result.points && result.points.length !== 0) {
+                this.logger.debug('read returned', result.points.length, 'points');
+                this.enqueue(result.points);
+            } else {
+                this.logger.debug('read returned no points');
+            }
+
+            this.readState = result.state;
+
+            // Check if we are being torn down.
+            if (this.inTeardown) {
+                return;
+            }
+
+            // The adapter indicates that it's completed reading all points up
+            // to the given end time by returning the ending time in readEnd.
+            //
+            // Notify the base class so it can emit any necessary ticks and
+            // check if we're all done, or call us back when it is time to
+            // reschedule.
+            //
+            // Otherwise we need to reschedule another call.
+            if (result.readEnd) {
+                this.advance(result.readEnd);
+            } else {
+                this.moreToRead = true;
```

Comment:
Oh ho, it's in `advance`. I wonder if it would be possible to have `emit_eof` in fewer places here, little hard to keep track of them all.